subprojects/gst-plugins-good/gst/rtp/gstrtph266depay.c

```diff
@@ -0,0 +1,674 @@
+      /* *INDENT-OFF* */
+      S = (payload[0] & 0b10000000) == 0b10000000;
+      E = (payload[0] & 0b01000000) == 0b01000000;
+      P = (payload[0] & 0b00100000) == 0b00100000;
+      FUType = payload[0] & 0b00011111;
+      /* *INDENT-ON* */
+
+      GST_DEBUG_OBJECT (rtph266depay,
+          "FU header with S %d, E %d, P %d FUType %d", S, E, P, FUType);
+
+      seqnum = gst_rtp_buffer_get_seq (rtp);
+
+      if (S) {                  // Start of NAL unit.
+        /* If a new FU unit started, while still processing an older one.
+         * Assume that the remote payloader is buggy (doesn't set the end
+         * bit) and send out what we've gathered thusfar */
+        if (G_UNLIKELY (rtph266depay->current_fu_type != 0)) {
+          gst_rtp_base_depayload_delayed (depayload);
+          gst_rtp_h266_finish_fragmentation_unit (rtph266depay);
+        }
+
+        rtph266depay->current_fu_type = nal_unit_type;
+        rtph266depay->fu_timestamp = timestamp;
+        rtph266depay->last_fu_seqnum = seqnum;
+        // rtph265depay->wait_start = FALSE;
+
+        /* reconstruct NAL header */
+        nal_header =
+            (((guint16) FUType) << 3) | (((guint16) nuh_layer_id) << 8) |
+            (guint16) nuh_temporal_id_plus1;
+        // GST_MEMDUMP_OBJECT (rtph266depay, "nal_header", (guint8 *) &nal_header,
+        //     sizeof (nal_header));
+
+        /* go back one byte so we can copy the payload + two bytes more in the front which
+         * will be overwritten by the nal_header
+         */
+        payload -= 1;
+        payload_len += 1;
+
+        nalu_size = payload_len;
+        outsize = nalu_size + sizeof (sync_bytes);
+        outbuf = gst_buffer_new_and_alloc (outsize);
+
+        gst_buffer_map (outbuf, &map, GST_MAP_WRITE);
+        memcpy (map.data, sync_bytes, sizeof (sync_bytes));
+        memcpy (map.data + sizeof (sync_bytes), payload, nalu_size);
+        map.data[sizeof (sync_bytes)] = nal_header >> 8;
+        map.data[sizeof (sync_bytes) + 1] = nal_header & 0xff;
+        gst_buffer_unmap (outbuf, &map);
+
+      } else {
+        if (rtph266depay->current_fu_type == 0) {
+          /* previous FU packet missing start bit? */
+          g_critical ("missing FU start bit on an "
+              "earlier packet. Dropping.");
+          gst_rtp_base_depayload_flush (depayload, FALSE);
+          gst_adapter_clear (rtph266depay->adapter);
+          return NULL;
+        }
+        if (gst_rtp_buffer_compare_seqnum (rtph266depay->last_fu_seqnum,
+                seqnum) != 1) {
+          /* jump in sequence numbers within an FU is cause for discarding */
+          g_critical ("Jump in sequence numbers from "
+              "%u to %u within Fragmentation Unit. Data was lost, dropping "
+              "stored.", rtph266depay->last_fu_seqnum, seqnum);
+          gst_rtp_base_depayload_flush (depayload, FALSE);
+          gst_adapter_clear (rtph266depay->adapter);
+          return NULL;
```

Comment:
Are we liberating all resources here?

Reply:
I was thinking about the `payload` and `rtp` variables, but as I mentioned in https://github.com/fluendo/gstreamer/pull/43#discussion_r1858812321 I no longer think there is a leak here.